File: linda-presentation/src/main/java/it/unibo/coordination/linda/presentation/LogicMatchDeserializer.java
package it.unibo.coordination.linda.presentation;

import com.fasterxml.jackson.databind.ObjectMapper;
import it.unibo.coordination.linda.logic.LogicMatch;
import it.unibo.coordination.linda.logic.LogicTemplate;
import it.unibo.coordination.linda.logic.LogicTuple;

import java.util.List;
import java.util.Map;

public class LogicMatchDeserializer extends DynamicDeserializer<LogicMatch> {

    public LogicMatchDeserializer(MIMETypes mimeType, ObjectMapper mapper) {
        super(LogicMatch.class, mimeType, mapper);
    }

    @Override
    public LogicMatch fromDynamicObject(Object dynamicObject) {
        if(dynamicObject instanceof List) {
            if(((List) dynamicObject).size() == 1) {
                dynamicObject = ((List) dynamicObject).get(0);
            }
        }
        if (dynamicObject instanceof Map) {
            final var dynamicMap = (Map<String, ?>) dynamicObject;

            if (dynamicMap.containsKey("template")) {
                final LogicTemplate template = getDeserializer(LogicTemplate.class).fromDynamicObject(dynamicMap.get("template"));
                final Object tupleObject;

                if ((tupleObject = dynamicMap.get("tuple")) != null) {
                    final LogicTuple tuple = getDeserializer(LogicTuple.class).fromDynamicObject(tupleObject);

                    return template.matchWith(tuple);
                }

                return LogicMatch.failed(template);
            }
        }
        throw new IllegalArgumentException("Cannot read " + getSupportedMIMEType());
    }
}

File: linda-logic-presentation/src/main/kotlin/it/unibo/coordination/linda/logic/LogicMatchDeserializer.kt
package it.unibo.coordination.linda.logic

import com.fasterxml.jackson.databind.ObjectMapper
import it.unibo.presentation.DynamicDeserializer
import it.unibo.presentation.MIMETypes

internal class LogicMatchDeserializer(mimeType: MIMETypes, mapper: ObjectMapper)
    : DynamicDeserializer<LogicMatch>(LogicMatch::class.java, mimeType, mapper) {

    @Suppress("NAME_SHADOWING")
    override fun fromDynamicObject(dynamicObject: Any): LogicMatch {
        var dynamicObject: Any? = dynamicObject
        if (dynamicObject is List<*>) {
            if (dynamicObject.size == 1) {
                dynamicObject = dynamicObject[0]
            }
        }
        if (dynamicObject is Map<*, *>) {

            val dynamicMap = dynamicObject
            if (dynamicMap.containsKey("template")) {
                val templateObj = dynamicMap["template"]
                        ?: error("Missing template field in parsing a ${LogicMatch::class.simpleName}")
                val template = Presentation.deserializerOf(LogicTemplate::class.java, supportedMIMEType)
                        .fromDynamicObject(templateObj)
                val tupleObject = dynamicMap["tuple"]
                if (tupleObject != null) {
                    val tuple = Presentation.deserializerOf(LogicTuple::class.java, supportedMIMEType)
                            .fromDynamicObject(tupleObject)
                    return template.matchWith(tuple)
                }
                return LogicMatch.failed(template)
            }
        }
        throw IllegalArgumentException("Cannot read $supportedMIMEType")
    }
}